apps/emqx/src/emqx_router_helper.erl

```diff
@@ -192,5 +112,37 @@ code_change(_OldVsn, State, _Extra) ->
 stats_fun() ->
     emqx_stats:setstat('topics.count', 'topics.max', emqx_router:stats(n_routes)).
 
+%% This is just to avoid cleaning up routes twice in case the node is down,
+%% as we can expect {mria, down}, {node, down} under normal circumstances
+%%  when a replicant node is being stopped.
+maybe_drain_nodedown_msg(DownNode) when DownNode =/= node() ->
+    receive
+        {membership, {_, down, DownNode}} -> ok
+    after ?DRAIN_TIMEOUT -> ok
+    end;
+maybe_drain_nodedown_msg(_) ->
+    ok.
+
 cleanup_routes(Node) ->
-    emqx_router:cleanup_routes(Node).
+    case mria_rlog:role() of
+        core ->
+            global:trans(
+                {?LOCK(Node), self()},
+                fun() -> emqx_router:cleanup_routes(Node) end,
+                %% Need to acquire the lock on replicants too,
+                %% since replicants participate in locking (see init/1).
+                [node() | nodes()],
+                %% It's important to do a limited number of retries (not infinity).
+                %% Otherwise, if a down node comes back in a short time
+                %% and wins the lock in init/1,
+                %% emqx_router_helepr on another node can desparetly retry it indefinitely.
+                2
+            );
+        replicant when Node =:= node() ->
+            %% Mnesia is down locally, but emqx_router_helper is still running and holds the lock,
```

Comment:
This whole concept `effectively` treats any crashes too seriously...
if `mria_membership` crashes on a replicant:
1. emqx_router_helper on this replicant node will delete the lock.
2. another node will receive `mria, down` event, acquire a lock and cleanup the routes
However, mria_membership proc can be restarted and the replicant will continue its operation, as though nothing happen...

Probably it's better to do routes cleanups only when the node is really down but not as soon as Mria (or even Mnesia is down)? :thinking:

Reply:
Changed `{mria, down}` events to be sent only when `mria_membership` terminates with `shutdown` reason, so that if it crashes and restarts due to some errors, it doesn't cause any cleanups. 
If Mria is shutdown on a remote node, it makes sense to cleanup routes even if that node keeps being connected at Erlang distr level..